lumicks/pylake/kymotracker/kymotracker.py

```diff
@@ -45,88 +20,80 @@ def track_greedy(
-        in.
+        Expected diffusion constant (default: 0.0). This parameter will influence whether a peak in
+        the next frame will be connected to this one. Increasing this value will make the algorithm
+        allow more positional variation in.
     sigma_cutoff : float
-        Sets at how many standard deviations from the expected trajectory a particle no longer belongs to this trace.
-        Lower values result in traces being more stringent in terms of continuing (default: 2.0).
+        Sets at how many standard deviations from the expected trajectory a particle no longer
+        belongs to this trace. Lower values result in traces being more stringent in terms of
+        continuing (default: 2.0).
     rect : tuple of two pixel coordinates
         Only perform tracking over a subset of the image. Pixel coordinates should be given as:
         ((min_time, min_coord), (max_time, max_coord)).
 
     References
     ----------
-    [1] Sbalzarini, I. F., & Koumoutsakos, P. (2005). Feature point tracking and trajectory analysis for video
-    imaging in cell biology. Journal of structural biology, 151(2), 182-195.
-    [2] Mangeol, P., Prevo, B., & Peterman, E. J. (2016). KymographClear and KymographDirect: two tools for the
-    automated quantitative analysis of molecular and cellular dynamics using kymographs. Molecular biology of the
-    cell, 27(12), 1948-1957.
+    [1] Sbalzarini, I. F., & Koumoutsakos, P. (2005). Feature point tracking and trajectory analysis
+    for video imaging in cell biology. Journal of structural biology, 151(2), 182-195.
+    [2] Mangeol, P., Prevo, B., & Peterman, E. J. (2016). KymographClear and KymographDirect: two
+    tools for the automated quantitative analysis of molecular and cellular dynamics using
+    kymographs. Molecular biology of the cell, 27(12), 1948-1957.
     """
-    data_selection = _get_rect(data, rect) if rect else data
-
-    sigma = sigma if sigma else 0.5 * line_width
-    coordinates, time_points = peak_estimate(
-        data_selection, np.ceil(0.5 * line_width), pixel_threshold
-    )
-    if len(coordinates) == 0:
-        return []
-
-    peaks = KymoPeaks(
-        *refine_peak_based_on_moment(
-            data_selection, coordinates, time_points, np.ceil(0.5 * line_width)
-        )
-    )
-    peaks = merge_close_peaks(peaks, np.ceil(0.5 * line_width))
-    lines = points_to_line_segments(
-        peaks,
-        kymo_score(vel=vel, sigma=sigma, diffusion=diffusion),
-        window=window,
-        sigma_cutoff=sigma_cutoff,
-    )
-
-    lines = [KymoLine(line.time_idx, line.coordinate_idx, data) for line in lines]
-
-    return KymoLineGroup(
-        [line.with_offset(rect[0][0], rect[0][1]) for line in lines] if rect else lines
+    return _track_greedy(
+        kymograph._get_channel(channel),
+        line_width,
+        pixel_threshold,
+        window,
+        sigma,
+        vel,
+        diffusion,
+        sigma_cutoff,
+        rect,
     )
```

Comment:
I'm not sure that the private API has a purpose here. I know that we talked about having a low-level private API and a high-level public API, but this isn't really what I had in mind. I figured that the private one would be uncalibrated and it would allow you to do unit tests more easily (more data without mocking). The high-level public API would add the calibration.

But this isn't really what ended up here. The private one does pretty much the same thing. Only one argument is different. I don't think that there is any advantage to doing this. It just seems like a lot of code duplication now.

Sorry if I wasn't clear when we talked about this way back. I didn't mean to imply that this should be split into two levels just for the sake of it. It's only useful if it helps the code be clearer (two nicely composed functions are clearer than one giant one) and the tests easier to write and understand (not as many things to set up). This didn't end up being either of those.

I think it should be reworked in one direction or the other. Doesn't really matter if it goes back or is split up better. Just as long as it doesn't stay in a not-so-useful middle.

Reply:
This wasn't done with the intention of making a distinction between low level and high level API. But it's true, I clearly didn't indicate well enough what I need/use these functions for.

The original design was to grab a `CalibratedKymographChannel` from whatever object the tracker widget got (through `_get_channel()`) and try to not tie the tracker directly to `Kymo`, but only use things that were in the specified interface `CalibratedKymographChannel` (to have no direct coupling between `Kymo` and the widget).

This had as extra benefit that when tracking multiple subsections of a Kymo (which is a feature), they can all hold a reference to the same `CalibratedKymographChannel`.

The alternative is that we just take a `Kymo` and create a new `CalibratedKymographChannel` for every set of lines and store both the `Kymo` and the `color_channel` in the widget and reconstruct a `CalibratedKymographChannel` every time we track a section of the image.

If you prefer the latter, I can change it around. I don't feel very strongly about it; since we could reintroduce it in the future.

Reply:
Alright. It has been axed.

The price to pay was a bit more coupling between `Kymo` and the actual widget and construction of a `CalibratedKymographChannel` for every tracking operation (but this cost is negligible compared to the actual tracking). Not the end of the world, especially now that we have the factories in `Kymo` (it may be more worthwhile to turn material coming from a WIT stack into a `Kymo` now rather than have a direct route to `CalibratedKymographChannel`; but let's see about that when we get there).